include/aspect/particle/property/cpo.h

```diff
@@ -0,0 +1,361 @@
+                                const std::pair<std::vector<double>, std::vector<Tensor<2,3> > > &derivatives,
+                                std::vector<double> &previous_volume_fraction_derivatives,
+                                std::vector<Tensor<2,3> > &previous_a_cosine_matrices_derivatives,
+                                const double strain_rate_second_invariant,
+                                const double dt) const;
+
+
+          /**
+           * A interface function which delegates the computation the derivatives
+           * to specialized functions.
+           */
+          std::pair<std::vector<double>, std::vector<Tensor<2,3> > >
+          compute_derivatives(const std::vector<double> &volume_fractions,
+                              const std::vector<Tensor<2,3> > &a_cosine_matrices,
+                              const SymmetricTensor<2,3> &strain_rate_nondimensional,
+                              const Tensor<2,3> &velocity_gradient_tensor_nondimensional,
+                              const double volume_fraction_mineral,
+                              const std::array<double,4> &ref_resolved_shear_stress) const;
+
+          /**
+           * Computes the update for the property rotating passively with the particle.
+           */
+          std::pair<std::vector<double>, std::vector<Tensor<2,3> > >
+          compute_derivatives_spin_tensor(const std::vector<double> &volume_fractions,
+                                          const std::vector<Tensor<2,3> > &a_cosine_matrices,
+                                          const SymmetricTensor<2,3> &strain_rate_nondimensional,
+                                          const Tensor<2,3> &velocity_gradient_tensor_nondimensional,
+                                          const double volume_fraction_mineral,
+                                          const std::array<double,4> &ref_resolved_shear_stress) const;
+
+
+          /**
+           * Declare the parameters this class takes through input files.
+           */
+          static
+          void
+          declare_parameters (ParameterHandler &prm);
+
+          /**
+           * Read the parameters this class declares from the parameter file.
+           */
+          void
+          parse_parameters (ParameterHandler &prm) override;
+
+          /**
+           * Return the number of grains per particle
+           */
+          static
+          unsigned int
+          get_number_of_grains();
+
+          /**
+           * Return the number of minerals per particle
+           */
+          static
+          unsigned int
+          get_number_of_minerals();
+
+        private:
+          // helpful shortcuts
+          double rad_to_degree = 180.0/M_PI;
+          double degree_to_rad = M_PI/180.0;
+
+          /**
+           * Random number generator used for initalization of particles
+           */
+          mutable boost::lagged_fibonacci44497 random_number_generator;
```

Comment:
Is it important to use this generator?  We usually either use the following generator or call a distribution directly:
```
    boost::mt19937                     rng(seed);
    boost::random::uniform_real_distribution<> uniform_distribution(-1, 1);
```

Reply:
Not a strong reason as far as I can remember (I probably just looked up the options and thought bigger cycle is better). In terms of speed it probably won't matter too much since since it is only used for initialization of the particle orientations.

Do you feel strongly about it? Because I am a bit hesitant to change it because I will probably have to update all the values in the unit test :P